### ISSUE TITLE
Support adding or removing dependencies

### ISSUE THREAD

**Issue by lnicola** [rust-lang/rust-analyzer#16443]
[`panicbit.cargo`](https://marketplace.visualstudio.com/items?itemName=panicbit.cargo) has some commands to add dependencies, maybe we should add something similar so people don't hate us after #16416.

**Comment by lnicola** [comment]
I got this working:

![image](https://github.com/rust-lang/rust-analyzer/assets/308347/03baa328-0cd4-4290-bd08-09d6827dcc44)

![image](https://github.com/rust-lang/rust-analyzer/assets/308347/4e62f170-0ebb-4f0e-b984-e6f9b1cca200)

But it pains me to bring back all that network and proxy stuff so I'm not sure we want it :sweat_smile:.

**Comment by alibektas** [comment]
A question at this point : Although definitely not the same in terms of user experience but am I right in assuming that we will be able to let user add and remove dependencies once we have `Cargo.toml` support? I expect user type a crate's name an suggestion fetched from "cargo search" populating suggestions.

**Comment by lnicola** [comment]
TIL `cargo search`. I guess it means we could do this without bringing in the network stuff in the VS Code extension, and even run it server-side?

But yes, it would also be nice to have some kind of completion support. Not sure how triggering them would work, though.

**Comment by lnicola** [comment]
Another way to offer this functionality would be via a fix-it: you type `use serde::Serialize`, get a diagnostic on `serde`, choose a quick fix, and a menu pops up with the crate search results. I think we can do that, right?

**Comment by Veykril** [comment]
Well we don't have name resolution diagnostics yet 😅 though for imports they shouldn't be too difficult to implement. Otherwise yes I think so?

**Comment by lnicola** [comment]
Yeah, I was wondering about the UI parts (opening the listbox after choosing the fix). I thought we have an assist that worked like that, but I don't remember which one.

**Comment by Veykril** [comment]
All the grouping assists do that (via an lsp extension iirc)

**Comment by panicbit** [comment]
Hey there! Would love if `cargo-edit` functionality was implemented in r-a directly!
I'm also happy to deprecate the addon and redirect to `r-a`!

**Comment by panicbit** [comment]
[Deprecation of `pancbit.cargo` is requested](https://github.com/microsoft/vscode-discussions/discussions/1)

**Comment by TethysSvensson** [comment]
Can we un-deprecate `panicbit.cargo` until these commands have been implemented in rust-analyzer?

**Comment by TethysSvensson** [comment]
@panicbit Is there any chance that you might be interested in bringing back your extension until this is supported by rust-analyzer?

**Comment by panicbit** [comment]
I'd be willing to work on releasing a version with cargo check/build functionality removed (leaving only `cargo add` and `cargo rm`).

**Comment by panicbit** [comment]
I have updated the extension, but we'll have to wait until it gets undeprecated again ([see request here](https://github.com/microsoft/vscode-discussions/discussions/1#discussioncomment-9817587)).

As a temporary workaround, you can download the extension from the marketplace and then drag it into the extension list in vscode:
https://marketplace.visualstudio.com/items?itemName=panicbit.cargo

**Comment by TethysSvensson** [comment]
For anyone else looking for a quick work-around: Pasting `ext install panicbit.cargo` in the Quick Open menu (Ctrl+P) also worked for me.

**Comment by sanmai-NL** [comment]
@lnicola Do you have a hyperlink to the relevant Issue(s) for rust-analyzer for us to read up on and track?

**Comment by lnicola** [comment]
@sanmai-NL not really. https://github.com/rust-lang/rust-analyzer/issues/3565 is a prerequisite for anything involving `Cargo.toml` on the server side, like support for completions. Given that, we could use the local index, but that would only suggest crates and versions that are already present in there. I don't think we want to include an HTTP client and look stuff up on crates.io. `cargo search` might also be an option, but that's a bit ugly.

We can also do this on the client-side, only for Code (like in my screenshot [at the top](https://github.com/rust-lang/rust-analyzer/issues/16443#issuecomment-1919813302)). But that would exclude other clients, of course.

As for the quick-fix route, I think we have a diagnostic, so we could, in principle, run `cargo search` and offer a fix.

**Comment by redactedontop** [comment]
@lnicola Could the taplo crate help?

**Comment by lnicola** [comment]
We know about `taplo` (and there's an open issue for real TOML support), but it's not really needed for this feature if we can use `cargo add`.

**Comment by vlovich** [comment]
FWIW I have two minor annoyances with panicbit's approach is:

1. Can't specify dev/build dependencies. Would be nice if it asked me.
2. I have a workspace that manages most of the dependencies - it would be nice to have some kind of option whereby I can add it to the workspace & then have it added into my project as workspce = true.
3. I like to keep my dependencies sorted to minimize merge conflicts but cargo add just blindly appends to the list of dependencies.

**Comment by lnicola** [comment]
> I like to keep my dependencies sorted to minimize merge conflicts but cargo add just blindly appends to the list of dependencies.

No, `cargo add` has been defaulting to sorted order for years, as long as the list is already sorted.

**Comment by blinsay** [comment]
This would be *amazing* with workspace support. My wishlist is:

1) The ability to add a dependency to the workspace's root Cargo.toml
2) Some kind of auto-complete-y action for adding a dependency to a workspace crate if the dependency is already included in the workspace root.

I find myself running into this scenario constantly in workspaces. Having support for these things in RA would take away a significant fraction of the pain I have there.
